File: concepts/07-graphs/node/edgeList.js
/**
 * Edge List Graph Representation in Node.js
 */

class EdgeListGraph {
    /**
     * Graph using Edge List representation
     */
    constructor(directed = false) {
        this.edges = [];
        this.vertices = new Set();
        this.directed = directed;
    }

    /**
     * Add edge - O(1)
     */
    addEdge(u, v, weight = 1) {
        this.edges.push([u, v, weight]);
        this.vertices.add(u);
        this.vertices.add(v);

        if (!this.directed) {
            this.edges.push([v, u, weight]);
        }
    }

    /**
     * Remove edge - O(E)
     */
    removeEdge(u, v) {
        if (this.directed) {
            this.edges = this.edges.filter(([a, b]) => !(a === u && b === v));
        } else {
            this.edges = this.edges.filter(
                ([a, b]) => !((a === u && b === v) || (a === v && b === u))
            );
        }
    }

    /**
     * Get all edges - O(1)
     */
    getEdges() {
        return [...this.edges];
    }

    /**
     * Get all vertices - O(V)
     */
    getVertices() {
        return Array.from(this.vertices);
    }

    /**
     * Get neighbors of vertex - O(E)
     */
    getNeighbors(vertex) {
        const neighbors = [];
        for (const [u, v, weight] of this.edges) {
            if (u === vertex) {
                neighbors.push([v, weight]);
            }
        }
        return neighbors;
    }

    /**
     * Check if edge exists - O(E)
     */
    hasEdge(u, v) {
        return this.edges.some(([a, b]) => a === u && b === v);
    }

    /**
     * Get edge weight - O(E)
     */
    getEdgeWeight(u, v) {
        for (const [a, b, weight] of this.edges) {
            if (a === u && b === v) {
                return weight;
            }
        }
        return null;
    }

    /**
     * Convert to adjacency list - O(E)
     */
    toAdjacencyList() {
        const graph = {};
        for (const [u, v, weight] of this.edges) {
            if (!graph[u]) {
                graph[u] = [];
            }
            graph[u].push([v, weight]);
        }
        return graph;
    }

    /**
     * Convert to adjacency matrix - O(V² + E)
     */
    toAdjacencyMatrix(numVertices = null) {
        if (numVertices === null) {
            numVertices = this.vertices.size;
        }

        const matrix = Array(numVertices).fill().map(() => Array(numVertices).fill(0));
        const vertexMap = {};
        const sortedVertices = Array.from(this.vertices).sort((a, b) => a - b);
        sortedVertices.forEach((v, i) => {
            vertexMap[v] = i;
        });

        for (const [u, v, weight] of this.edges) {
            const i = vertexMap[u];
            const j = vertexMap[v];
            if (i !== undefined && j !== undefined) {
                matrix[i][j] = weight;
            }
        }

        return { matrix, vertexMap };
    }

    /**
     * Kruskal's MST using edge list (already sorted by weight)
     */
    kruskalMST() {
        // Sort edges by weight
        const sortedEdges = [...this.edges].sort((a, b) => a[2] - b[2]);

        class UnionFind {
            constructor(vertices) {
                this.parent = {};
                this.rank = {};
                for (const v of vertices) {
                    this.parent[v] = v;
                    this.rank[v] = 0;
                }
            }

            find(x) {
                if (this.parent[x] !== x) {
                    this.parent[x] = this.find(this.parent[x]);
                }
                return this.parent[x];
            }

            union(x, y) {
                const px = this.find(x);
                const py = this.find(y);
                if (px === py) {
                    return false;
                }

                if (this.rank[px] < this.rank[py]) {
                    [px, py] = [py, px];
                }

                this.parent[py] = px;
                if (this.rank[px] === this.rank[py]) {
                    this.rank[px]++;
                }
                return true;
            }
        }

        const uf = new UnionFind(this.vertices);
        const mst = [];

        for (const [u, v, weight] of sortedEdges) {
            if (uf.union(u, v)) {
                mst.push([u, v, weight]);
                if (mst.length === this.vertices.size - 1) {
                    break;
                }
            }
        }

        return mst;
    }
}

// Example usage
if (require.main === module) {
    // Undirected Graph
    console.log("--- Edge List Graph (Undirected) ---");
    const graph = new EdgeListGraph(false);
    graph.addEdge(0, 1, 4);
    graph.addEdge(0, 2, 1);
    graph.addEdge(1, 2, 2);
    graph.addEdge(1, 3, 5);
    graph.addEdge(2, 3, 8);

    console.log("Edges:", graph.getEdges());
    console.log("Vertices:", graph.getVertices());
    console.log("Neighbors of 1:", graph.getNeighbors(1));
    console.log("Has edge (0,1):", graph.hasEdge(0, 1));
    console.log("Weight of (0,1):", graph.getEdgeWeight(0, 1));

    // Convert to adjacency list
    console.log("\nAdjacency List:");
    const adjList = graph.toAdjacencyList();
    for (const [vertex, neighbors] of Object.entries(adjList)) {
        console.log(`${vertex}:`, neighbors);
    }

    // Kruskal's MST
    console.log("\nKruskal's MST:");
    const mst = graph.kruskalMST();
    console.log("MST Edges:", mst);
    console.log("Total Weight:", mst.reduce((sum, [, , w]) => sum + w, 0));

    // Directed Graph
    console.log("\n--- Edge List Graph (Directed) ---");
    const dgraph = new EdgeListGraph(true);
    dgraph.addEdge(0, 1, 1);
    dgraph.addEdge(1, 2, 2);
    dgraph.addEdge(2, 0, 3);
    console.log("Edges:", dgraph.getEdges());
}

module.exports = { EdgeListGraph };

File: concepts/07-graphs/python/edge_list.py
"""
Edge List Graph Representation in Python
"""

class EdgeListGraph:
    """Graph using Edge List representation"""
    
    def __init__(self, directed=False):
        self.edges = []
        self.vertices = set()
        self.directed = directed
    
    def add_edge(self, u, v, weight=1):
        """Add edge - O(1)"""
        self.edges.append((u, v, weight))
        self.vertices.add(u)
        self.vertices.add(v)
        
        if not self.directed:
            self.edges.append((v, u, weight))
    
    def remove_edge(self, u, v):
        """Remove edge - O(E)"""
        if self.directed:
            self.edges = [(a, b, w) for a, b, w in self.edges if not (a == u and b == v)]
        else:
            self.edges = [(a, b, w) for a, b, w in self.edges 
                         if not ((a == u and b == v) or (a == v and b == u))]
    
    def get_edges(self):
        """Get all edges - O(1)"""
        return self.edges.copy()
    
    def get_vertices(self):
        """Get all vertices - O(V)"""
        return list(self.vertices)
    
    def get_neighbors(self, vertex):
        """Get neighbors of vertex - O(E)"""
        neighbors = []
        for u, v, weight in self.edges:
            if u == vertex:
                neighbors.append((v, weight))
        return neighbors
    
    def has_edge(self, u, v):
        """Check if edge exists - O(E)"""
        for a, b, _ in self.edges:
            if a == u and b == v:
                return True
        return False
    
    def get_edge_weight(self, u, v):
        """Get edge weight - O(E)"""
        for a, b, weight in self.edges:
            if a == u and b == v:
                return weight
        return None
    
    def to_adjacency_list(self):
        """Convert to adjacency list - O(E)"""
        graph = {}
        for u, v, weight in self.edges:
            if u not in graph:
                graph[u] = []
            graph[u].append((v, weight))
        return graph
    
    def to_adjacency_matrix(self, num_vertices=None):
        """Convert to adjacency matrix - O(V² + E)"""
        if num_vertices is None:
            num_vertices = len(self.vertices)
        
        matrix = [[0] * num_vertices for _ in range(num_vertices)]
        vertex_map = {v: i for i, v in enumerate(sorted(self.vertices))}
        
        for u, v, weight in self.edges:
            i = vertex_map.get(u)
            j = vertex_map.get(v)
            if i is not None and j is not None:
                matrix[i][j] = weight
        
        return matrix, vertex_map
    
    def kruskal_mst(self):
        """Kruskal's MST using edge list (already sorted by weight)"""
        # Sort edges by weight
        sorted_edges = sorted(self.edges, key=lambda x: x[2])
        
        class UnionFind:
            def __init__(self, vertices):
                self.parent = {v: v for v in vertices}
                self.rank = {v: 0 for v in vertices}
            
            def find(self, x):
                if self.parent[x] != x:
                    self.parent[x] = self.find(self.parent[x])
                return self.parent[x]
            
            def union(self, x, y):
                px, py = self.find(x), self.find(y)
                if px == py:
                    return False
                
                if self.rank[px] < self.rank[py]:
                    px, py = py, px
                
                self.parent[py] = px
                if self.rank[px] == self.rank[py]:
                    self.rank[px] += 1
                return True
        
        uf = UnionFind(self.vertices)
        mst = []
        
        for u, v, weight in sorted_edges:
            if uf.union(u, v):
                mst.append((u, v, weight))
                if len(mst) == len(self.vertices) - 1:
                    break
        
        return mst


# Example usage
if __name__ == "__main__":
    # Undirected Graph
    print("--- Edge List Graph (Undirected) ---")
    graph = EdgeListGraph(directed=False)
    graph.add_edge(0, 1, 4)
    graph.add_edge(0, 2, 1)
    graph.add_edge(1, 2, 2)
    graph.add_edge(1, 3, 5)
    graph.add_edge(2, 3, 8)
    
    print("Edges:", graph.get_edges())
    print("Vertices:", graph.get_vertices())
    print("Neighbors of 1:", graph.get_neighbors(1))
    print("Has edge (0,1):", graph.has_edge(0, 1))
    print("Weight of (0,1):", graph.get_edge_weight(0, 1))
    
    # Convert to adjacency list
    print("\nAdjacency List:")
    adj_list = graph.to_adjacency_list()
    for vertex, neighbors in adj_list.items():
        print(f"{vertex}: {neighbors}")
    
    # Kruskal's MST
    print("\nKruskal's MST:")
    mst = graph.kruskal_mst()
    print("MST Edges:", mst)
    print("Total Weight:", sum(weight for _, _, weight in mst))
    
    # Directed Graph
    print("\n--- Edge List Graph (Directed) ---")
    dgraph = EdgeListGraph(directed=True)
    dgraph.add_edge(0, 1, 1)
    dgraph.add_edge(1, 2, 2)
    dgraph.add_edge(2, 0, 3)
    print("Edges:", dgraph.get_edges())

